(DNM) Wip d4n filter test program

samarahu opened the pull request:
Currently, all tests are consistently passing.

mosayyebzadeh committed:
rgw/d4n: squashing all commits related to initial implementation
of write-back cache for non-multipart objects (writing objects to
cache and then to backend store in the cleaning process):

1. combining write cache with latest D4N code
2. cleaning the code
3. removing some bugs on bigger objects
4. updating iterate function to check the dirty flag
5. Updating write cache based on the pull request comments.
Read process needs to be updated based on write process. It needs to check where is the data and if it is dirty or clean.
If it is in the cache and dirty, we need to put D_ in the oid of the object before reading it from cache.
If it is clean, there is nothing to do.
6. updating flush functions and comments.

Signed-off-by: mosayyebzadeh <mosayyeb@bu.edu>

samarahu committed:
rgw/d4n: Add directory probing to read workflow

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: implementation of caching head in read workflow.

modifications in ReadOp::prepare() method of the d4n filter driver
to cache the head object.

modification in get_obj_attrs to read from cache or backend store.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing all commits related to caching head in the
write-back workflow, modifying set_obj_attrs(), get_obj_attrs() and
delete_obj_attrs based on the cached head and modifying the cleaning
method to use a min-heap data structure for storing dirty objects only.

1. rgw/d4n: implementation for caching head object in write-back workflow.
2. rgw/d4n: modifications to get write back cache working after cleaning
process.
3. rgw/d4n: modifications for eviction of dirty blocks.
4. rgw/d4n: modifications include adding a heap of dirty objects
which has objects ordered by their creation time and the top
element of which is fetched in the cleaning method, processed
and deleted in a loop.
5. rgw/d4n: changing the format of cached blocks to
bucket_name_version_object_name_ofs_len, to avoid checks
for versioned and non-versioned objects.
6. rgw/d4n: modifications to set_obj_attrs(), modify_obj_attrs()
and delete_obj_attrs() to check if the head object exists in a cache,
else direct the calls to backend store.
7. rgw/d4n: handling version in case of bucket versioning being suspended
while writing the object.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: squashing the following commits for miscellaneous fixes
to policy driver, redis driver, directory and filter driver.

1. d4n/directory: Use boost::split for simpler code
2. rgw: Lower log levels for failures in D4N and redis cache files
3. rgw: Add dpp and logs to directory, cache, and policy
4. rgw: Reduce Redis calls and fix workflow
5. qa/d4n: Remove D4N task and add S3 user creation to workunit driver script
6. d4n: Use Redis transactions to serialize consecutive requests for safe
data handling and faster completion
7. d4n/directory: Remove boost lexical_cast calls
8. rgw/d4n: Add return values to error logs
9. rgw/d4n: Change directory hostsList to use `unordered_set`
10. d4n/filter: Simplify logic for storing block in `handle_data`
11. rgw/policy: Properly delete `LFUDAEntry` instances
12. rgw/d4n: Add support for `dirty` block metadata, `check_bool` for consistent values, and fix directory updates in `cleanup` method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/test: squashing the following commits related to
D4N test failures.

1 d4n: Add `rename` method and unit test and update D4N unit tests
2. d4n/test: Fix linking issue

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: fix to correctly populate rgw_obj of GetObjectState in
RGWRados, in case ReadOp::prepare() reads the head object from
the cache.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: implementation of copyObject.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n: Add filter `delete_obj` implementation and complete GET support

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: miscellaneous corrections/additions:
1. storing objects in directory using their oid, so that the version
is included.
2. making sure that the head block corresponds to latest
version in the block directory.
3. add a directory entry for head block for every version
in case of a versioned bucket.
4. Populating hostsList correctly for blocks and objects.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: Squash the following commits related to directory, eviction and filter driver logs:
1. d4n/directory: Remove `blockHosts` and change `objHosts` to store the hosts for the block
2. d4n/directory: Remove `dirtyBlock` metadata and use the cacheObj's `dirty` flag in block logic
3. rgw/d4n: This commit introduces the following changes:
  a. Fix `get_victim_block` method to correctly handle version
  b. Return `0` instead of error for non-error circumstances in `eviction`
  c. Switch `increase` to `decrease` for heap after localWeight update in
   `eviction`
  d. Update filter writer logs
4. d4n/filter: Make minor adjustments

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: adding perfcounters for d4n cache hits, misses and evictions (LFUDA).

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing the following commits that fix s3 test
failures.

1. rgw/d4n: fixing logical errors for ranged requests in iterate(), aligning
start chunk/block to beginning of the chunk/block boundary.
2. cache head object in write path in case write-cache
is disabled - for maintaining cache consistensy
3. delete dirty blocks of older versions of a non-versioned
bucket as dirty blocks are not evicted, and older versions won't
be written to the disk also.
4. remove dirty objects from the dirty objects data structure,
in case the bucket it belongs to has been deleted.
5. rgw/d4n: fixing compilation issue after rebasing with main.
6. Handle multipart objects correctly in copy object
7. Fix more ranged request errors.
8. handling part_num correctly for multipart and non-multipart objects
9. handling obect name starting with '_'.
10. bug in range request for sending back data to the client, after reading from
backend store.
11. rgw/d4n: code to populate RGW_ATTR_CRYPT_PARTS attr
in d4n filter.
12. rgw/d4n: fix for 'if-match' and 'if-nomatch' for
put object and copy object.
13. rgw/cache: Account for slashes in object name
14. d4n/filter: Handle request headers during read if `next->prepare` isn't called
15. d4n/filter: Fix overwriting of object size and instance

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/filter: Implement `get_bucket` and `load_bucket`

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: removing code that writes object entry to the
object directory in redis

Objects were being written to the directory for bucket listing
but another data structure needs to be used that allows for
ordered listing indexed by a bucket name/ id, hence removing
the entries that do not serve this purpose (they also add to the
number of calls to the redis server).

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

sinashan committed:
bucket listing with directory

Signed-off-by: sinashan <sinatak1373@live.com>

pritha-srivastava committed:
rgw/d4n: miscellaneous updates:

1. use bucket id instead of name in the cache block key
as this will be helpful to differentiate between objects
belonging to buckets that have been deleted and recreated.
2. add a helper method to derive prefix for cache blocks.
3. comments and code cleanup, renaming of policy methods
related to adding and erasing dirty objects to and from
in-memory data structures.
4. rgw/d4n: reading in chunks of rgw_max_chunk_size instead
of rgw_obj_max_req_size since data is being written in rgw_max_chunk_size
chunks in the write back cache.
5. rgw/d4n: correctly set the version for objects
with 'null' versionId in case of a write workflow.
The version stored in the block directory was 'null',
but that does not differentiate two different versions
of objects with versionId set to 'null', hence differentiating
between objects with 'null' versionId by deriving and storing
a non-null version in the block directory, and based on this
version the name of the object in the cache is derived.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing the following commits for restoring
in-memory data for LFUDA policy and dirty objects from
cache on disk

1. rgw/d4n: restore in memory data structure for dirty
objects using the xattrs in the head block of an object
in the cache.
2. rgw/d4n: replacing string cache attr names with constexpr char.
3. rgw/d4n: restore LFUDA policy data from cache on disk.
4. rgw/d4n: correcting the key used while updating the LFUDA data structure
for the head object block in the read path.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: handling special characters in object names such that
parsing file names while restoring data from cache does not result
in errors.

1. Introduce a new delimiter # instead of _ for concatening bucket_id,
objectname etc to create a key, because "_" is a valid char in object names
and is not encoded and results in incorrect values while parsing filenames
for restorin data.
2. url encode the bucket id, version and object name before all are concatenated
to form a key
3. url_decode the bucket id, version and object name after parsing the filename
from cache while restoring data back.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: squashing commits for supporting delete object
for unversioned buckets.

1. d4n/directory: Update logging
2. d4n/filter: Update `set_head_obj_dir_entry` to handle null versioning and update `prevVersion` for versioned head objects
3. rgw/d4n: Simplify code to only support unversioned bucket deletes

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/cache: implementation for organizing data in ssd cache
into bucket directory, object directory and then versions
representing file names.

This commit for extracts bucket id, object name, version,
offset, and length from input key and organizes data into
directories - with bucket id, object name being used to
create directories and filenames within them using version
and offset and length.

Data can be restored back to in memory LFUDA and dirty object
data structure once rgw is restarted.

Updating unit tests.
test/rgw: Correct RedisDriver keys and add checks for cache blocks after
write

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/directory: support for ordered set in both ObjectDirectory
and BlockDirectory.

Adds methods to add, list a range and remove members from a
redis ordered set ordered by score passed in. Also adds unit
test cases to test the methods.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: support for dirty objects in delete object method.

1. ordered set to maintain versions of a dirty object
2. creation of delete marker in case of a simple delete request
3. deletion of a specific version from the ordered set
4. cleaning method deletes from ordered set for dirty objects
5. use of redis atomicity constructs wherever needed

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing following commits that add support for
WATCH, EXEC, INCR, MULTI and DISCARD redis commands.

1. d4n/directory: support for watch, exec and incr methods in Block
and Object Directory.
2. d4n/directory: support for redis MULTI and DISCARD command,
adding a test case to exercise usage of MULTI.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: squashing commits related to handling of
clean (non-dirty) objects for delete_obj method.

1. rgw/d4n: miscellaneous fixes that include modifying
set_head_obj_dir_entry() and delete_obj method to handle
clean objects only.
2. rgw/d4n: changes to design when write cache is turned off:
a. do not store a hash entry for latest version for versioned
buckets, the latest version will always be fetched from backend store,
to maintain correct order of versions during a delete object request.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: adding 'HSET' and 'DEL' to transactions to make the
operations atomic while modifying the latest and null hash entry
in case of non-versioned buckets.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: squashing following commits related
to lazy deletion - delete dirty objects data
blocks in cleaning method and non-dirty objects
data blocks in eviction method

1. rgw/d4n: Implement lazy deletion
2. rgw/d4n: cleaning method now supports delete markers
also (both versioned and null). The delete markers
are written correctly to the backend store.
Also modifying the description of rgw_d4n_cache_cleaning_interval
to explicitly state that the duration is in seconds.
3. rgw/d4n: do not call invalidate_dirty_object in case
of a simple delete request for dirty objects belonging
to a versioned bucket.
In this case, a delete marker needs to be created instead of
invalidating/deleting an object.
4. rgw/d4n: Update lazy deletion in policy

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/cache: Remove `del` implementation from RedisDriver

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: squashing following commits to address concurrency in
filter driver and ssd backed cache.

1. rgw/d4n: addressing concurrency issues by adding a refcount
to each block.
Blocks with positive refcounts are pinned (similar to dirty blocks)
and not eligible for eviction or deletion. Updating unit tests also.
2. rgw/cache: addressing concurrency issues while directories creation,
deletion, updating xattrs, get_attr and put.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -520,8 +520,12 @@ int D4NFilterObject::calculate_version(const DoutPrefixProvider* dpp, optional_y
     } else {
-      ldpp_dout(dpp, 0) << __func__ << " Failed to find id tag" << dendl;
-      return -ENOENT;
+      auto attrs = this->get_attrs();
```
why is invoking get_attrs() needed? Also what if after calling get_attrs() we still do not find the ID_TAG? Whenever you make such a change, please elaborate why these changes are needed in the commit message. One thing to keep in mind is that the code has worked in the past without calling get_attrs() here, so have we introduced a regression somewhere along the way? If we have, then we must fix the regression instead of changing a method that is invoked from a lot of other methods which might introduce other bugs.

pritha-srivastava replied:
Infact I went and checked and every method that invokes calculate_version() also invoked get_attrs() before invoking it.

samarahu replied:
Yes, I meant to add a comment in regards to this so apologies for not initially doing so. The reason I added this change is because, as the object is being written during the `complete` call in the unit test program, `calculate_version()` fails to find the ID_TAG with the get_attr() call. Originally, I was also trying to set that attribute explicitly with set_attrs(), but this did not work either. I asked Dan about this and he said get_attrs() and set_attrs() doesn't set anything in the backend store, so `calculate_version()` will call get_attr() and this method will check the `has_attrs` attribute, which ends up false as a result. This causes issues with the rest of the test due to key parsing inconsistencies, so he said to use get_attrs() to extract attrs from the cached map in the Object class instead of the store. Please let me know if swapping to a `get_attrs()` call here causes any issues conceptually. I can rerun local tests to validate it with the simpler workflows we've established thus far.

pritha-srivastava replied:
I have already mentioned that all methods calling calculate_version() invoke get_attrs before calculate_version is invoked, so I am not sure why the extra call is needed. If you are seeing this issue only with db store as the backend, then the test program needs to be changed and not the code here.

pritha-srivastava replied:
Does db store provide ID_TAG as an attribute, or do you need to set it as part of attrs that you are sending in the complete step?

pritha-srivastava replied:
https://github.com/ceph/ceph/pull/56336/files#diff-5cf52d0baced08513b69bbb3cf4a29682289fce58451b9567c6d2e86ff080bceR1156 is the line no, where get_attrs() is invoked in ReadOp::prepare() and then calculate_version is invoked at https://github.com/ceph/ceph/pull/56336/files#diff-5cf52d0baced08513b69bbb3cf4a29682289fce58451b9567c6d2e86ff080bceR1158. You must check why the attr is not available when get_attrs is called in 1156.

pritha-srivastava replied:
You probably need to pass in the attrs to the calculate_version() method - please check this and modify the method accordingly.

samarahu replied:
dbstore does not provide an ID_TAG, so I must write that in the test explicitly. Passing the `attrs` in as a parameter to calculate_version works, however, so I have made this change. Please let me know if this is what you had in mind.

pritha-srivastava replied:
this is correct, however add a comment that dbstore does not have an ID tag, therefore you are passing it as an attr since D4N cache backend versioning relies on it.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,441 @@
+
+  buffer::list bl;
+  bl.append("test_version");
+  rgw::sal::Attrs attrs{{RGW_ATTR_ID_TAG, bl}};
```
why is the ID_TAG being passed in as an attribute? does dbstore not return ID_TAG as part of the attrs when read from the backend.

samarahu replied:
I was having trouble with retrieving the ID_TAG during initial debugging, so I explicitly wrote it to the attrs to follow the value more closely. Since I changed `calculate_version()`, I will remove this and see if dbstore provides one instead.

samarahu replied:
dbstore does not provide this attribute, so I must set it explicitly

pritha-srivastava replied:
ok, see comment above

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,441 @@
+    // Check directory values
+    boost::system::error_code ec;
+    request req;
+    req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0"); // TODO: add null head object check as well, or is that being removed?
```
null head object will not be removed.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,441 @@
+  io.run();
+}
+
+TEST_F(D4NWriteCacheFixture, PutObject)
```
why is a separate fixture needed for a Write cache?

samarahu replied:
The initialization is different since I must change the config value but I can change the value at the beginning of each test instead of creating a separate fixture entirely if you believe that is the cleaner approach

pritha-srivastava replied:
If it is easier for you to get the tests done with separate fixtures, then go for it. We can combine them later when we have the time.

samarahu committed:
rgw/d4n: Remove dirty prefix and make xattr instead

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: implementation of list objects for a bucket
(ordered listing) and list-object-versions.

Changes include the addition of an ordered set for
each bucket that stores the object, addition of etag
and object size to CacheObj structure needed for listing
objects, and addition of a test case to see the usage of
pipelining in Redis.  Objects are fetched from the cache
and the backend store and are then merged, before returning
the final list.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n/directory: Add more error handling to update_field

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -523,18 +524,15 @@ void D4NFilterObject::set_attrs_from_obj_state(const DoutPrefixProvider* dpp, op
 }
 
-int D4NFilterObject::calculate_version(const DoutPrefixProvider* dpp, optional_yield y, std::string& version)
+int D4NFilterObject::calculate_version(const DoutPrefixProvider* dpp, optional_yield y, std::string& version, rgw::sal::Attrs& attrs)
```
Supply attributes to method rather than retrieving them internally

samarahu commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2819 @@
+      EXPECT_EQ((int)std::get<2>(resp).value(), 0);
+    }
+
+    /* TODO: Eviction cycle to delete cache blocks
```
TODO: add eviction cycle testing

samarahu commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2819 @@
+    put_version_enabled_object(testName, instance, yield);
+    put_version_suspended_object(testName, yield);
+
+    // TODO: No unversioned head object set in redis, is this intended?
```
TODO: Do unversioned head objects get set in the directory for versioned objects?

pritha-srivastava replied:
Yes, now we fetch latest version from backend store for versioned clean objects due to complexities in maintaining versions between the cache and backend store.

samarahu replied:
Is "yes" in response to the GitHub comment or the code comment? They are phrased differently so I wanted to confirm

pritha-srivastava replied:
I think both mean the same, don't they? If not please clarify. I meant that clean objects belonging to versioned buckets do not have a latest and null hash entry in redis.

samarahu commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2819 @@
+    EXPECT_EQ(read_op_suspended->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_suspended->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
+
+    // TODO: data block set without version in key, is this intended?
```
TODO:  Is it expected for objects with versioning suspended to not have their version ("null") in the directory entry key for the data block?

pritha-srivastava replied:
if the version is "null" the oid is the name of the object, the oid does not include "null" as the version, so this is expected.

samarahu commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2819 @@
+      EXPECT_EQ((int)std::get<4>(resp).value(), 0);
+    }
+
+    /* TODO: Eviction cycle to delete cache blocks
```
TODO: eviction cycle testing

samarahu commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2819 @@
+      EXPECT_EQ(std::get<0>(resp).value(), "0");
+    }
+
+    // TODO: check if written to backend
```
TODO: check if object is available through backend

mkogan1 committed:
rgw/d4n: fix valgrind Invalid Memory Read & Write

Signed-off-by: Mark Kogan <mkogan@redhat.com>

mkogan1 committed:
rgw/d4n: fix output of binary data to text log and logs cleanup

the RGW_ATTR_ACL attr logging requires special handling

Signed-off-by: Mark Kogan <mkogan@redhat.com>

pritha-srivastava committed:
rgw/d4n: removing calls to redis MULTI, DISCARD, WATCH,
UNWATCH and EXEC since other options for Redis transactions
are being explored.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -523,18 +524,15 @@ void D4NFilterObject::set_attrs_from_obj_state(const DoutPrefixProvider* dpp, op
-    if (this->get_attr(RGW_ATTR_ID_TAG, bl)) {
-      version = bl.c_str();
+    bufferlist bl = attrs[RGW_ATTR_ID_TAG];
+    version = bl.to_str();
```
the old code had version set to bl.c_str(), is there a reason for changing that to bl.to_str()?

samarahu replied:
No, I will update this

pritha-srivastava replied:
ok

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -523,18 +524,15 @@ void D4NFilterObject::set_attrs_from_obj_state(const DoutPrefixProvider* dpp, op
+    if (!version.empty()) {
       ldpp_dout(dpp, 20) << __func__ << " id tag version is: " << version << dendl;
-    } else {
-      ldpp_dout(dpp, 0) << __func__ << " Failed to find id tag" << dendl;
```
you can leave the else here, just remove the return ENOENT part.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    }
+
+    void put_version_suspended_object(std::string name, boost::asio::yield_context yield) {
+      testBucket->get_info().flags |= BUCKET_VERSIONS_SUSPENDED;
```
Is there no method to enable/suspend versioning (instead of directly accessing the flags?)

samarahu replied:
This is how versioning is modified in the posix test. I asked Dan about how to change bucket versioning but this is the only way he knew how. I couldn't find other examples, either

pritha-srivastava replied:
Ok

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    std::unique_ptr<rgw::sal::Object> destObj = testBucket->get_object(rgw_obj_key(destName));
+    EXPECT_NE(destObj.get(), nullptr);
+
+    int ret = obj->copy_object(acl_owner,
```
where is obj declared?

samarahu replied:
It is part of the test fixture

pritha-srivastava replied:
Ok

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    EXPECT_EQ((int)std::get<1>(resp).value(), 1);
+    EXPECT_EQ(std::get<2>(resp).value().size(), 10);
+    EXPECT_EQ(std::get<3>(resp).value().size(), 10);
+
```
In all tests of CopyObj, where you test based on different flags for xattrs, you should validate whether the xattrs have been modified/not modified based on the flag passed in, like get the xattrs from the head block and see if the xattrs set are according to the expected values.

samarahu replied:
Shouldn't this only apply to the write cache copy object tests since, in the read cache, the attributes are overwritten by the `dest_object->get_attrs()` call?

pritha-srivastava replied:
Well, for a read-cache, the attrs manipulation is performed by the backend store when next->copy_object is invoked, so dest_object->get_attrs() should still have the correct attrs. You can leave the check for a read cache since the attrs calculation is done by the backend store and not by the filter driver.

samarahu replied:
By leaving the check, do you mean not including it since it is not done on the filter level?

pritha-srivastava replied:
yes do not include it.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("FLUSHALL");
```
what is the need for a FLUSHALL here?

samarahu replied:
I add `FLUSHALL` calls at the end of each test when I am finished using redis in order to set up for the next test. I will simply move this call to the setup method

pritha-srivastava replied:
that looks better

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    EXPECT_EQ(testData, "test data");
+    testFile.close();
+
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + version, err), true);  
```
where is the version declared and defined?

samarahu replied:
The version is declared and defined on line 803

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      EXPECT_EQ((int)std::get<0>(resp).value(), 1);
+      EXPECT_EQ(std::get<1>(resp).value().size(), 10);
+
+      std::error_code err;
```
similar comment as above - wherever you test copyObject with different xattrs flag, you can add checks to see if the result is what you expect them to be.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+	       false,
+	       nullptr,
+	       nullptr,
+	       rgw::sal::ATTRSMOD_NONE,
```
shouldn't the flag be rgw::sal::ATTRSMOD_MERGE, based on the name of the test?

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    std::string oid = version + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + oid, err), true);     
+    oid = instance + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + instance, err), true);     
```
I think you intend to check `oid` here in place of `instance`

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0");
```
you don't need a check for hash entry for latest version, as in the read cache for version enabled buckets we do not save the latest entry in the directory, as explained earlier.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+
+  io.run();
+}
+
```
For read cache, you can add tests where you copy an object, then read it and check whether the cache contents are correct, for both version enabled and non-versioned buckets.

samarahu replied:
DBStore's `copy_object` is not implemented and simply returns 0. As a result, the object is not written to the backend during the copy_object call and the subsequent read op will fail.

pritha-srivastava committed:
rgw/d4n: adding check for null target_obj in
get_obj_attrs() method to prevent crashes.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+  env->cct->_conf->d4n_writecache_enabled = true;
+  env->cct->_conf->rgw_d4n_cache_cleaning_interval = 0.5;
+  const std::string testName = "PutObjectWrite";
+  const std::string bucketName = "/tmp/d4n_filter_tests/dbstore-default_ns.1";
```
can you explain here why you have prefixed the bucket name with the directory path?

samarahu replied:
I did this because the filter sets dirty entries with the bucket id rather than the bucket name during several calls, so when the cleaning method executes, it searches for the DBStore bucket path.

pritha-srivastava replied:
all across the d4n filter we bucket_id. I still do not undertsand why you need to prefix the directory path. Please explain with line numbers pointing to code. we already have the cache path also set in a config option, and we derive the bucket id based on the bucket name.

samarahu replied:
https://github.com/samarahu/ceph/blob/wip-d4n-filter-test/src/rgw/driver/d4n/d4n_policy.cc#L670

Here, I printed the bucket name and bucket ID and saw that they hold different values during the test.

pritha-srivastava replied:
yes, the bucket id is different from the bucket name. what problem are you facing or trying to solve here by adding the prefix?

samarahu replied:
I am setting the bucket name and ID to be equivalent so the filter and cleaning method use the same key when updating D4N values

pritha-srivastava replied:
That's what I don't understand - why is this needed - please explain in more detail. With the actual code, we do not have any such requirement. And what D4N values are you referring to here?

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+
+    // Ensure object is written to backend
+    auto next = dynamic_cast<rgw::sal::FilterObject*>(obj.get())->get_next();
+    EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
```
great work!

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    std::ifstream testFile; 
+
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + destName + "/" + version, err), true); // TODO: version here isn't original version, is this expected? 
+    std::string oid = version + "#0#" + std::to_string(ofs);
```
same comment for copy object as read test cases - validate the xattrs.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    std::string testData; 
+    std::ifstream testFile; 
+
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + destName + "/" + version, err), true); // TODO: version here isn't dest version, is this expected? 
```
Yes, for write-back cache, the version is calculated internally and is not based on ID tag since the object does not exist in the backend store and ID tag is available only when the object is in backend store and is used in the read cache. Hence version here will not be dest_version set to ID tag.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    std::string testData; 
+    std::ifstream testFile; 
+
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + destName + "/" + version, err), true); // TODO: version here isn't dest version, is this expected? 
```
same explanation as previous test case

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      req.push("EXISTS", bucketName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", bucketName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("EXISTS", bucketName + "__:null_"  + TEST_OBJ + testName + "_0_0");
+      req.push("ZRANGE", bucketName + "_" + TEST_OBJ + testName, "0", "10000000000");
```
ordered set stores elements in ascending order, and we store versions based on the creation time, so the latest version will be saved last in the set. You can use zrevrange to get the versions and the top element must be the latest version which is "null" and then the next version will be the instance version. If you use zrange, the order will be reversed.

samarahu replied:
While using `ZRANGE`, I've had instances where the second value is the instance value rather than "null". I'm not sure why this happens but because it doesn't occur when the redis `monitor` command is being used, I assume it has something to do with the operations overlapping during runtime. As a result, I added logic that checks for either "null" or the instance value for the first set entry and checks for the opposite value for the second set entry.

pritha-srivastava replied:
I suggest you use zrevrange, for the reason I explained above, and you can use 0 as start and -1 as stop for listing all members

pritha-srivastava replied:
actually "null" version also gets stored in the ordered set in addition to the version - this is needed for list object versions command.

samarahu replied:
@pritha-srivastava I tried using `ZREVRANGE` but the results are still not consistent and return in different orders during each run

pritha-srivastava replied:
@samarahu : I am not sure why that happens, you need to debug that. once you run the test, you can use the command line to see what values the ordered set holds.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      EXPECT_EQ((int)std::get<3>(resp).value(), 1);
+      EXPECT_EQ((int)std::get<4>(resp).value(), 1);
+      EXPECT_EQ((int)std::get<5>(resp).value(), 1);
+      EXPECT_EQ(std::get<6>(resp).value()[0], instance);
```
You can add a check for "null" version also

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op_enabled(objEnabled->get_read_op());
+    EXPECT_EQ(read_op_enabled->prepare(optional_yield{yield}, env->dpp), 0);
+    EXPECT_EQ(read_op_enabled->iterate(env->dpp, 0, ofs, nullptr, optional_yield{yield}), 0);
```
This test must validate the data returned against the expected data, so that we are sure that data being read from the cache is correct.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    rgw_placement_rule placement;
+    ceph::real_time mtime;
+    buffer::list bl;
+    bl.append("dest_object_version");
```
again here dest_version will not taken from the ID tag as explained in one of the previous tests.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      req.push("EXISTS", TEST_BUCKET + testName + "__:" + instance + "_" + destNameEnabled + "_0_9");
+      req.push("ZRANGE", TEST_BUCKET + testName + "_" + destNameEnabled, "0", "10000000000");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:" + instance + "_" + destNameEnabled + "_0_0");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:" + instance + "_" + destNameEnabled + "_0_0");
```
Is there a difference between line 2275 and 2274?

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + destNameSuspended + "_0_9");
+      req.push("ZRANGE", TEST_BUCKET + testName + "_" + destNameSuspended, "0", "10000000000");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:null_" + destNameSuspended + "_0_0");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:null_" + destNameSuspended + "_0_0");
```
again, is there a difference between 2357 and 2356 line nos.?

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    std::string instance; 
+    put_version_enabled_object(testName, instance, yield);
+
+    RGWEnv rgw_env;
```
again for all copyobject tests, validation of xattrs is a must.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      req.push("EXISTS", TEST_BUCKET + testName + "__:" + instance + "_" + destNameEnabled + "_0_9");
+      req.push("ZRANGE", TEST_BUCKET + testName + "_" + destNameEnabled, "0", "10000000000");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:" + instance + "_" + destNameEnabled + "_0_0");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:" + instance + "_" + destNameEnabled + "_0_0");
```
check line 2472 and 2473

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + destNameSuspended + "_0_9");
+      req.push("ZRANGE", TEST_BUCKET + testName + "_" + destNameSuspended, "0", "10000000000");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:null_" + destNameSuspended + "_0_0");
+      req.push("HGETALL", TEST_BUCKET + testName + "__:null_" + destNameSuspended + "_0_0");
```
check line 2554 and 2555

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+    rgw_zone_id zone;
+    rgw_placement_rule placement;
+    ceph::real_time mtime;
+    buffer::list bl;
```
same comments as above copyobject tests and check for duplicate line nos also

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,2810 @@
+
+  io.run();
+}
+
```
very good test coverage! I want you to add a couple of more tests for the writeback cache to test delete object for versioned bucket:
1. Upload 2 versions, issue a simple delete request - check whether that is a delete marker , after cleaning verify that the cached block is removed
2. Upload 2 versions, delete the latest instance and then verify the directory contents (using zrevrange) that the correct instance is promoted as the correct one. You can also validate that the deleted instance is not written to the backend store during the cleaning cycle. and that the cached blocks get deleted.
3. Upload 2 versions. Wait for the cleaning cycle to complete. Then issue a simple delete, then make sure a delete marker is created (and this will be created in the backend store)
4. Upload 2 versions. Wait for cleaning cycle to complete. Then delete an instance and check if it gets correctly deleted from the backend store.
5. Upload object to a non-versioned bucket. Then change the version of the bucket to enabled and upload a version of the same object, and then to suspended and upload a version of the object. Let the cleaning cycle complete. Verify that the order of the versions are correct in the backend store (Null followed by instance).

samarahu replied:
For the first workflow after a simple delete, why would any cache blocks be removed? Shouldn't the workflow complete like described in #3?

samarahu replied:
For the last test, what method should I use to return the versions in order?

pritha-srivastava replied:
list-object-versions

pritha-srivastava replied:
@samarahu : can you list the names of the new tests here?

mkogan1 committed:
rgw/d4n: write to the cache with O_SYNC flag by default

add conf to allow modifying cache write 'man 2 open' flags
with O_SYNC enabled by default.

performance impact measurment -
workload is PUT of 100K 5KB objects

- with O_SYNC flag: 3890 OP/s,
  iostat IO util% range while workload is running: ~45%-49%

- without O_SYNC flag: 4511 OP/s,
  iostat IO util% range while workload is running: ~10%-25%

Signed-off-by: Mark Kogan <mkogan@redhat.com>

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+    conn->async_exec(req, resp, yield[ec]);
+
+    ASSERT_EQ((bool)ec, false);
+    deleteMarker = std::get<0>(resp).value();
```
will this return true or false?

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+    deleteMarker = std::get<0>(resp).value();
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + deleteMarker, err), false);  
```
deleteMarker should evaluate to 'true' hence the file won't be found, we do create a head block with version as the file name and attach an xattr to it to check if the file is a placeholder for a delete marker or not. I think this check needs to be modified

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(optional_yield{yield});
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + deleteMarker, err), false);  
```
here you just need to do a simple get and see if the result returned is a delete marker, and check if the existing head block in the cache has been marked clean

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+      EXPECT_EQ(std::get<0>(resp).value().size(), 1);
+      EXPECT_EQ(std::get<0>(resp).value()[0], instances[0]);
+    }
+
```
along with these checks you also need to perform a simple 'get' and confirm whether the versionId returned is the correct next latest version.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+    // First instance should still be written to backend
+    {
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
```
where are you verifying the version? are you setting the instance value and then calling load_obj_state? and if the object is found then the version exists? Think of a way to verify that the latest version written to the backend is the correct one.

samarahu replied:
Yes, I'm setting the instance value to the first instance that wsan't deleted, and using this to get the next object, which should represent that instance in the backend. Since I can load the object and explicitly check if it exists in the line below, I believe that proves that particular instance is available in the backend.

pritha-srivastava replied:
If there are two instances and you set an instance (even the second one which is not the latest) and call load_obj_state(), even then you would be able to see the object, so that doesn't check for the correct latest version, hence I asked you to do a simple get and check the latest version. And I can see that you have added code to verify the latest version.

samarahu replied:
Yes, I added the simple get call for this purpose but should I remove the calls to `next->load_obj_state` then? On line 3224, I did set the instance to the second instance, which has been deleted, and the `next->load_obj_state` call returns -2 to represent this.

pritha-srivastava replied:
No what you have done is correct.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+    
+    for (int i = 0; i <= 1; ++i) {
+      std::string instance;
+      put_version_enabled_object(testName, instance, yield); // Upload two versions 
```
the comment says upload two versions - does put_version_enabled_object upload two versions? I see that the method first enables, uploads, suspends and then again uploads an object. You can think of a name that signifies this.

samarahu replied:
The method itself only uploads one version but it is in a loop here for a double upload.

pritha-srivastava replied:
oh yes! Ignore my comment

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(optional_yield{yield});
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + instances[1], err), true);  
```
You must add a check to see if the blocks are clean.

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+      auto next = dynamic_cast<rgw::sal::FilterObject*>(objEnabled.get())->get_next();
+      EXPECT_EQ(next->load_obj_state(env->dpp, optional_yield{yield}), 0);
+      EXPECT_EQ(next->exists(), 1);
+    }
```
similar comment - try to verify what is the latest version id, is there a way to create a read_op of the backend and then call its prepare() and after that check the instance returned

pritha-srivastava commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+  }, rethrow);
+
+  io.reset();
+  io.run_for(std::chrono::seconds(2));
```
why is this needed?

samarahu commented on src/test/rgw/test_d4n_filter.cc:
```diff
@@ -0,0 +1,3449 @@
+    d4nFilter->get_policy_driver()->get_cache_policy()->update_dirty_object(env->dpp, key, "", true, objEnabled->get_accounted_size(), creationTime, 
+                                                         std::get<rgw_user>(objEnabled->get_bucket()->get_owner()), etag, 
+                                                         objEnabled->get_bucket()->get_name(), objEnabled->get_bucket()->get_bucket_id(), 
+                                                         objEnabled->get_key(), rgw::d4n::REFCOUNT_NOOP, optional_yield{yield});*/
```
@pritha-srivastava I've added a comment here since the cleaning method doesn't properly clean the head block for the delete marker in the cache. The DBStore delete_obj method seems to handle the call differently than RadosStore, so the cleaning process gets interrupted with a failure

pritha-srivastava replied:
what happens when delete_obj is called with del_op params, with version id set to the version id of the delete marker in the cache? If the delete marker is not created successfully in the backend store, what error is returned back?

samarahu replied:
The cleaning method is still interrupted with a `-ENOENT` return value from DBstore's `delete_obj` method. Setting the delete marker to be the object's instance supplies the op with a version id and as you can see in the logic for the `delete_obj` method, if a version id is supplied, `-ENOENT` is returned: https://github.com/samarahu/ceph/blob/wip-d4n-filter-test/src/rgw/driver/dbstore/common/dbstore.cc#L1879

samarahu committed:
rgw/d4n: Change score calculation

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: Update creationTime calculation to include decimal value

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/d4n: making LFUDAPolicy destructor lock free to prevent
deadlocks during rgw shutdown.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -67,8 +67,8 @@ int LFUDAPolicy::init(CephContext *cct, const DoutPrefixProvider* dpp, asio::io_
 
   driver = _driver;
   if (dpp->get_cct()->_conf->d4n_writecache_enabled) {
+    quit = false;
```
Without explicitly resetting the `quit` value here, it remains true in later tests and as a result, the cleaning method does not execute.

pritha-srivastava replied:
ok

samarahu committed:
test/d4n: Add D4N filter test program

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: Supply attrs to `calculate_version`

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
